File: src/llm.ts
import Anthropic from "@anthropic-ai/sdk";
import { GoogleGenAI } from "@google/genai";
import OpenAI from "openai";
import { settings } from "./config.js";
import type { ModelInfo } from "./config.js";

/**
 * One streaming entry point over five LLM providers. Mirrors the Python
 * `stream_llm` generator: yields only non-empty text chunks, and throws on
 * an unrecognized provider.
 */
export async function* streamLlm(
  prompt: string,
  modelInfo: Pick<ModelInfo, "provider" | "value" | "base_url">,
): AsyncGenerator<string> {
  const { provider, value: modelName } = modelInfo;

  if (provider === "openai" || provider === "xai") {
    const apiKey = provider === "xai" ? settings.XAI_API_KEY : settings.OPENAI_API_KEY;
    const client = new OpenAI({ apiKey, baseURL: modelInfo.base_url });
    const stream = await client.chat.completions.create({
      model: modelName,
      messages: [{ role: "user", content: prompt }],
      temperature: 0.6,
      stream: true,
    });
    for await (const chunk of stream) {
      const text = chunk.choices?.[0]?.delta?.content;
      if (text) yield text;
    }
  } else if (provider === "anthropic") {
    const client = new Anthropic({ apiKey: settings.ANTHROPIC_API_KEY });
    const stream = client.messages.stream({
      model: modelName,
      max_tokens: 8192,
      temperature: 0.6,
      messages: [{ role: "user", content: prompt }],
    });
    for await (const event of stream) {
      if (
        event.type === "content_block_delta" &&
        event.delta.type === "text_delta" &&
        event.delta.text
      ) {
        yield event.delta.text;
      }
    }
  } else if (provider === "gemini") {
    const client = new GoogleGenAI({ apiKey: settings.GOOGLE_API_KEY });
    const stream = await client.models.generateContentStream({
      model: modelName,
      contents: [{ role: "user", parts: [{ text: prompt }] }],
      config: { temperature: 0.6 },
    });
    for await (const chunk of stream) {
      const text = chunk.text;
      if (text) yield text;
    }
  } else if (provider === "openrouter") {
    const client = new OpenAI({
      apiKey: settings.OPENROUTER_API_KEY,
      baseURL: "https://openrouter.ai/api/v1",
    });
    // OpenRouter's `provider` routing hint isn't part of the OpenAI SDK's
    // types (it's an OpenRouter-specific body extension, equivalent to the
    // Python client's `extra_body`), so it's added via a typed cast.
    const body = {
      model: modelName,
      messages: [{ role: "user", content: prompt }],
      temperature: 0.6,
      stream: true,
      provider: { order: ["DeepInfra"] },
    } as OpenAI.Chat.Completions.ChatCompletionCreateParamsStreaming & {
      provider: { order: string[] };
    };
    const stream = await client.chat.completions.create(body);
    for await (const chunk of stream) {
      const text = chunk.choices?.[0]?.delta?.content;
      if (text) yield text;
    }
  } else {
    throw new Error(`Unknown provider: ${provider as string}`);
  }
}

File: yapzee_common/llm.py
import os
from openai import OpenAI
from anthropic import Anthropic
from google import genai
from google.genai import types as genai_types
from typing import Generator
from yapzee_common.config import settings

def stream_llm(prompt: str, model_info: dict) -> Generator[str, None, None]:
    provider = model_info["provider"]
    model_name = model_info["value"]
    
    if provider in ["openai", "xai"]:
        api_key = settings.XAI_API_KEY if provider == "xai" else settings.OPENAI_API_KEY
        client = OpenAI(api_key=api_key, base_url=model_info.get("base_url"))
        response = client.chat.completions.create(
            model=model_name,
            messages=[{"role": "user", "content": prompt}],
            temperature=0.6,
            stream=True
        )
        for chunk in response:
            if chunk.choices and chunk.choices[0].delta.content:
                yield chunk.choices[0].delta.content

    elif provider == "anthropic":
        client = Anthropic(api_key=settings.ANTHROPIC_API_KEY)
        with client.messages.stream(
            model=model_name,
            max_tokens=8192,
            temperature=0.6,
            messages=[{"role": "user", "content": prompt}]
        ) as stream:
            for text in stream.text_stream:
                if text:
                    yield text

    elif provider == "gemini":
        client = genai.Client(api_key=settings.GOOGLE_API_KEY)
        response = client.models.generate_content_stream(
            model=model_name,
            contents=[genai_types.Content(role="user", parts=[genai_types.Part(text=prompt)])],
            config=genai_types.GenerateContentConfig(temperature=0.6)
        )
        for chunk in response:
            if hasattr(chunk, "text") and chunk.text:
                yield chunk.text
    elif provider == "openrouter":
        client = OpenAI(api_key=settings.OPENROUTER_API_KEY, base_url="https://openrouter.ai/api/v1")
        response = client.chat.completions.create(
            model=model_name,
            messages=[{"role": "user", "content": prompt}],
            temperature=0.6,
            stream=True,
            extra_body={"provider": {"order": ["DeepInfra"]}}
        )
        for chunk in response:
            if chunk.choices and chunk.choices[0].delta.content:
                yield chunk.choices[0].delta.content

    else:
        raise ValueError(f"Unknown provider: {provider}")
